### PR TITLE
Fix unicode literal parsing to allow 5-6 digit `U+` sequences and disallow `<` in abbreviations

### LINKED ISSUE #91: Is there a sanctioned way to reference a code point with its official name?
The [README](https://github.com/rbuckton/grammarkdown/blob/main/README.md#character-literals-and-ranges) mentions two ways to reference a Unicode code point, but fails to adequately specify them:

<blockquote>
<ul>
<li>An abbreviation for a Unicode Code point, of the form <code>&lt;NBSP&gt;</code></li>
<li>A Unicode code point, of the form <code>U+00A0</code></li>
</ul>
</blockquote>

[grammarkdown.grammar](https://github.com/rbuckton/grammarkdown/blob/main/spec/grammarkdown.grammar) doesn't mention the latter at all, and [implicitly defines the former](https://github.com/rbuckton/grammarkdown/blob/2ecf056a1b774733afd73113533dfad9f7ad44c8/spec/grammarkdown.grammar#L55-L62) as one or more non-`<` non-`>` non-|LineTerminator| code points in between `<` and `>`. As for the implementation, [scanner.ts uses `scanString(CharacterCodes.GreaterThan, …)`](https://github.com/rbuckton/grammarkdown/blob/2ecf056a1b774733afd73113533dfad9f7ad44c8/src/scanner.ts#L416), which pays special attention only to line terminators and `>`—and in particular _allows_ `<` when represented as a character reference like `&lt;` in e.g.
```
Nonterminal :::
  &lt;foo&lt;bar&gt;
```
[scanner.ts also handles the second form](https://github.com/rbuckton/grammarkdown/blob/main/src/scanner.ts#L509-L518) upon encountering "U+" or "u+" followed by four hexadecimal digits (and notably _not_ working for supplementary-plane characters such as U+1D306 TETRAGRAM FOR CENTRE "𝌆").

----

This is relevant because I want to express a nonterminal like `<U+2212 MINUS SIGN>`, which is not clearly valid or invalid according to documentation here and accepted by ecma262 `build:spec` while being rejected by esmeta (cf. https://github.com/tc39/ecma262/pull/3098/commits/cc5e203926bd54dc3e296915df0f9ca8a2e947aa and https://github.com/tc39/ecma262/actions/runs/5270397258/jobs/9529840136?pr=3098 ).

Ideally, we'd end up with alignment between documentation and implementation on a form that represents a single code point in any Unicode plane by its hexadecimal value _plus_ descriptive explanatory text (generally its name in the Unicode Character Database), e.g.
<blockquote>
<p>A single Unicode code point may be specified using one of the following forms:</p>
<ul>
<li><code>U+</code> followed by four to six non-lowercase hexadecimal digits with no leading zeroes other than those necessary for padding to a minimum of four digits, in accordance with <a href="https://www.unicode.org/versions/Unicode15.0.0/appA.pdf">The Unicode Standard, Version 15.0.0, Appendix A, Notational Conventions</a> (i.e., matching Unicode extended BNF pattern <code>"U+" ( [1-9 A-F] | "10" )? H H H H</code> or regular expression pattern <code>^U[+]([1-9A-F]|10)?[0-9A-F]{4}$</code> as in <code>U+00A0</code> or <code>U+1D306</code>)</li>
<li>The preceding representation followed by a space and a printable ASCII prose explanation (such as a character name) free of <code>&lt;</code> and <code>></code> and line terminators, all wrapped in <code>&lt;</code> and <code>></code> (i.e., matching Unicode extended BNF pattern <code>"&lt;" "U+" ( [1-9 A-F] | "10" )? H H H H " " [\u0020-\u007E -- [&lt;>]]+ ">"</code> or regular expression pattern <code>^&lt;U[+]([1-9A-F]|10)?[0-9A-F]{4} [\x20-\x3b\x3d\x3f-\x7e]+>$</code> as in <code>&lt;U+2212 MINUS SIGN></code>)</li>
<li>An abbreviation defined somewhere outside the grammar as an ASCII identifier name (i.e., matching Unicode extended BNF pattern <code>[A-Z a-z _] [A-Z a-z _ 0-9]*</code> or regular expression pattern <code>^[A-Za-z_][A-Za-z_0-9]*$</code> as in <code>&lt;NBSP></code>)</li>
</blockquote>

### PR DESCRIPTION
**Pull request by rbuckton**
Fixes #91

### CODE REVIEW

**Comment by rbuckton** [comment]
This now contains all of the rules suggested in #91.

File: spec/grammarkdown.grammar [hunk 1]
```diff
@@ -44,26 +51,77 @@ Keyword :: one of
...
+	HexDigits HexDigit
+
 HexDigit :: one of
 	`0` `1` `2` `3` `4` `5` `6` `7` `8` `9` `a` `b` `c` `d` `e` `f` `A` `B` `C` `D` `E` `F`
 
+NonZeroHexDigit ::
+	HexDigit but not `0`
+
 UnicodeCharacterLiteral ::
 	`<` UnicodeCharacterLiteralChars `>`
```
Comment (gibson042) [inline spec/grammarkdown.grammar:70]: This is much more general than the "ASCII identifier name" mentioned in the [README](https://github.com/rbuckton/grammarkdown/tree/main#character-literals-and-ranges), and probably _too_ general given the inclusion of |LineContinuation|. Perhaps strike a middle ground by requiring ASCII for `<U+…>` in accordance with Unicode conventions but allowing arbitrary identifiers in other `<…>` instances?
```suggestion
PrintableASCIIChars ::
	PrintableASCIIChar
	PrintableASCIIChars PrintableASCIIChar

PrintableASCIIChar ::
	U+0020 through U+007F

UnicodeCharacterLiteral ::
	`<` IdentifierName `>`
	`<` `U+` UnicodeCharacterLiteralHexDigits U+0020 PrintableASCIIChars `>`
```

**Review by gibson042 — COMMENTED** [review]
Thanks!

File: spec/grammarkdown.grammar [hunk 2]
```diff
@@ -44,26 +51,77 @@ Keyword :: one of
...
+	`U+` UnicodeCharacterLiteralHexDigits
+	`u+` UnicodeCharacterLiteralHexDigits
```
[the remark this answers is not in the document]
Comment (rbuckton) [reply spec/grammarkdown.grammar:72]: I parse it, but print a warning as part of other diagnostics.

File: spec/grammarkdown.grammar (hunk 1 above)
Comment (rbuckton) [reply spec/grammarkdown.grammar:70]: I wouldn't worry too much about that file, I need to completely rewrite it

**Pull request merged**